### PR TITLE
add fixed/immutable parameter prefix

### PR DESCRIPTION
**Pull request by AndrewPickford**
This adds a parameter prefix ('=') that makes a parameter fixed or immutable, ie later merges on top of this parameter either generate an error or do not change the parameter (depending on an option setting [_ignore_merging_onto_fixed_]).

For example with:
```yaml
# nodes/node1.yml
classes:
- first
- second

# classes/first.yml
parameters:
 =A: 1

# classes/second.yml
parameters:
 A: 2
```

reclass would if _ignore_merging_onto_fixed_ was false raise an error on the attempt to change the value of A in second.yml or if _ignore_merging_onto_fixed_ was true the parameter A would be evaluated to 1 and the attempt to change A in second.yml would be ignored.

This is an alternative way of fixing a parameter value rather than repeatedly loading a class (#34).

One downside is that when a parameter is fixed and _ignore_merging_onto_fixed_ is true then the non change of the parameter is silent and makes understanding how a parameter gets it's final value more difficult. So this is a use only were necessary option.

There's also a change to the _wrap_dict, _wrap_list and _wrap_value methods in parameters.py that got convoluted with the fixed parameters. The patch removes the need to pass DictPath objects into the _wrap_dict, _wrap_list and _wrap_value methods. The DictPath objects were mostly redundant only being used if there was an error, instead try/except blocks are used to build DictPath objects only when an error occurs. This gets rid of about half of the calls to DictPath.\_\_init\_\_, giving a small speed up of a few percent.

### CODE REVIEW

**Comment by richardfelkl** [comment]
Seems like a good feature. It will be useful in some specific cases.

File: README-extentions.rst [hunk 1]
```diff
@@ -124,6 +124,55 @@ Referenced lists or dicts can now be overriden by None or empty type of dict, li
     three: ${one}
 
 
+Fixed/Immutable Parameters
```
Comment (epcim) [inline README-extentions.rst]: fixed -> (static, freeze, frozen) to express better what it does?

File: README-extentions.rst [hunk 2]
```diff
@@ -124,6 +124,55 @@ Referenced lists or dicts can now be overriden by None or empty type of dict, li
...
+    alpha:
+      one: 1
+
+if ``ignore_merging_onto_fixed`` is true
+
+Default value for ``ignore_merging_onto_fixed`` is False
+
+.. code-block:: yaml
+
+  ignore_merging_onto_fixed: False
```
Comment (epcim) [inline README-extentions.rst]: With more and more features comming I would rather prefer "switch" that will enable/disable that feature, assuming we will rename "fixed" to "constant":

constant_params_enabled: False

**Comment by epcim** [comment]
Approved. (Waiting whether you will rename the options/discussion, will be merged soon)

**Comment by AndrewPickford** [comment]
Agreed that constant would be better name than fixed. For the switch, this changes what happens when something is merged on top of a constant parameter, so:

constant_params_enabled: False

isn't right - I'd assume an option with that name selected if the '=' prefix had a special meaning or not. Instead how about:

strict_constant_params: False   or
strict_constant_params_enabled: False

although the enabled is a bit redundant. Then if strict is set an attempt to merge on top of  a constant parameter generates an error otherwise the merge is ignored and the parameter remains unchanged.

If that's OK, I can update the patch in a day or two.

**Comment by epcim** [comment]
Hi, sure _enabled is redundant.  (BTW _params should rather be "_parameters" for even better readability.

From my point `constant_parameters: False` fit's best, anyway nothing against `strict_constant_parameters: False`.  As author, choose what you prefer ;)

**Comment by AndrewPickford** [comment]
Added a new commit with the updated naming scheme, should be ready to be merged with develop.

**Comment by epcim** [comment]
Cool!

**Pull request merged**